src/main/scala/is/hail/expr/AST.scala

```diff
@@ -742,15 +742,22 @@ case class Apply(posn: Position, fn: String, args: Array[AST]) extends AST(posn,
       } yield ir.ApplyBinaryPrimOp(op, x, y)
     }
 
+  private def tryComparisonConversion: IndexedSeq[(Type, IR)] => Option[IR] = flatLift {
+    case IndexedSeq((xt, x), (yt, y)) => for {
+      op <- ir.ComparisonOp.fromStringAndTypes.lift((fn, xt, yt))
+    } yield ir.ApplyComparisonOp(op, x, y)
+  }
+
   private[this] def tryIRConversion(agg: Option[(String, String)]): ToIRErr[IR] =
     for {
       irArgs <- all(args.map(_.toIR(agg)))
       ir <- fromOption(
         this,
         s"no function or primop found for $fn",
         tryPrimOpConversion(args.map(_.`type`).zip(irArgs)).orElse(
+          tryComparisonConversion(args.map(_.`type`).zip(irArgs)).orElse(
           IRFunctionRegistry.lookupConversion(fn, irArgs.map(_.typ))
```

Comment:
We expect exactly one of these to match right? This is a good opportunity to create a utility function that asserts that behavior.

```scala
def exactlyOne(
  blame: IR,
  message: String,
  a: Option[T],
  b: Option[T],
  c: Option[T]
): ToIRErr[T] = (a, b, c) match {
  case (Some(x), None, None) => success(x)
  case (None, Some(x), None) => success(x)
  case (None, None, Some(x)) => success(x)
  case (x, y, z) => fail(blame, "expected exactly one of $x, $y, $z to be Some.", ToIRErr.getCaller())
}
```

I think the use site looks a little nicer too less `orElse` noise

```scala
    for {
      irArgs <- all(args.map(_.toIR(agg)))
      typedIRArgs = args.map(_.`type`).zip(irArgs)
      ir <- fromOption(
        this,
        s"no function or primop found for $fn",
        exactlyOneOf(
          tryPrimOpConversion(typedIRArgs),
          tryComparisonConversion(typedIRArgs),
          IRFunctionRegistry.lookupConversion(fn, irArgs.map(_.typ))
            .map { irf => irf(irArgs) })))
// ...
```

Also can we get away without `typedIRArgs`? the IR has the type now right?